File: 20. Graph - II/3. Graph/1. Bellmon Ford.py
# relaxation : n-1 times
# n = nodes

from collections import defaultdict

class Graph:
    def __init__(self):
        self.adjList = defaultdict(list)

    def addEdge(self, u, v, wt, direction):
        # direction = 1 -> undirected graph
		    # direction => 0 -> directed graph;
        self.adjList[u].append((v, wt))
        if direction:
            self.adjList[v].append((u, wt))

    def printAdjList(self):
        for i in self.adjList:
            print(i, "->", end=" ")
            for j in self.adjList[i]:
                print("(", j[0], ",", j[1], ")", end=" ")
            print()

    def bellmanFordAlgo(self, n, src):
        # assuming directed weightted graph
        dist = [float("inf")] * n
        dist[src] = 0

        # n-1 relaxation step
        for _ in range(n - 1):
            # for all edges
            for u in self.adjList:
                for v, wt in self.adjList[u]:
                    if dist[u] != float("inf") and dist[u] + wt < dist[v]:
                        dist[v] = dist[u] + wt

        # to check for -ve cycle
        negativeCycle = False
        for u in self.adjList:
            for v, wt in self.adjList[u]:
                if dist[u] != float("inf") and dist[u] + wt < dist[v]:
                    negativeCycle = True
                    break

        if negativeCycle:
            print("-ve cycle present")
        else:
            print("-ve cycle absent")

        print("Printing dist array:")
        for i in dist:
            print(i, end=" ")

g = Graph()

g.addEdge(0, 1, -1, False)
g.addEdge(0, 2, 4, False)
g.addEdge(1, 2, 3, False)
g.addEdge(1, 3, 2, False)
g.addEdge(1, 4, 2, False)
g.addEdge(3, 1, 1, False)
g.addEdge(3, 2, 5, False)
g.addEdge(4, 3, -3, False)

g.printAdjList()
g.bellmanFordAlgo(5, 0)

File: 20. Graph - II/3. Graph/classcode/1. Bellmon Ford.cpp
// relaxation : n-1 times
// n = nodes

#include <iostream>
#include<unordered_map>
#include<list>
#include<queue>
#include<stack>
#include<algorithm>
#include<set>
#include<limits.h>
using namespace std;

class graph {
	public:
	unordered_map<int, list<pair<int,int> > > adjList;

	void addEdge(int u, int v, int wt, bool direction) {
		//direction = 1 -> undirected graph
		//direction => 0 -> directed graph;
		adjList[u].push_back({v,wt});
		if(direction == 1) {
			adjList[v].push_back({u,wt});
		}
	}

	void printAdjList() {
		for(auto i: adjList) {
			cout << i.first <<"-> ";
			for(auto j: i.second) {
				cout << "(" << j.first<<", "<<j.second<<"), ";
			}
			cout << endl;
		}
	}

  	void bellmanFordAlgo(int n, int src) {
		//assuming directed weightted graph
		vector<int> dist(n, INT_MAX);
		dist[src] = 0;
		//n-1 relaxation step
		for(int i=0; i<n-1; i++) {
			//for all edges
			for(auto t: adjList) {
				for(auto nbr: t.second) {
					int u = t.first;
					int v = nbr.first;
					int wt = nbr.second;
					if(dist[u] != INT_MAX && dist[u] + wt < dist[v]) {
						dist[v] = dist[u] + wt;
					}
				}
			}
		} 

		//to check for -ve cycle
		bool negativeCycle = false;
		for(auto t: adjList) {
				for(auto nbr: t.second) {
					int u = t.first;
					int v = nbr.first;
					int wt = nbr.second;
					if(dist[u] != INT_MAX && dist[u] + wt < dist[v]) {
						negativeCycle = true;
						break;
					}
				}
			}

		if(negativeCycle == true) 
			cout << "-ve cycle present"<< endl;
		else
			cout << "-ve cycle absent" << endl;

		cout << "printing dist array: "<< endl;
		for(auto i : dist)
			cout << i << " ";
		
	}
};

int main() {
	graph g;

	g.addEdge(0,1,-1,0);
	g.addEdge(0,2,4,0);
	g.addEdge(1,2,3,0);

	g.addEdge(1,3,2,0);
	g.addEdge(1,4,2,0);

	g.addEdge(3,1,1,0);
	g.addEdge(3,2,5,0);
	g.addEdge(4,3,-3,0);

	g.printAdjList();
	g.bellmanFordAlgo(5, 0);

  	return 0;
}